Add "mark messages as unread" feature

gnprice opened the pull request:
This implements #5364. ~The branch is based on #5544.~ (→ #5544 is now merged; rebased on that.)

I've manually tested, and this worked fine in my testing.

Marked as draft because ~this version will not work properly when there are messages the action should apply to but which we don't have locally -- which can start happening when the number of applicable messages is more than about 50. For a robust implementation, we'll want the upcoming API feature discussed in zulip/zulip#22893. Once that's available (which is expected before the upcoming Zulip Server 6.0 release), we'll switch to using that.~ it relies on the draft new API endpoint from https://github.com/zulip/zulip/pull/23512, implementing zulip/zulip#22893.

Fixes: #5364

chrisbobbe reviewed (commented):
Thanks, glad to have this prototype! I've tested it too and it seems to work. See a few comments below.

The events sometimes seemed to take multiple seconds to arrive in the app, compared with a fraction of a second in the web app, which I had open simultaneously. (I added logging in `startEventPolling` just after the `await api.pollForEvents` line to see when the app received an event.) But I don't think that's related to these changes.

gnprice commented on src/action-sheets/index.js:
```diff
@@ -179,6 +184,31 @@ const markTopicAsRead = {
+        return;
+      }
+      const affectedIds = fetchedMessageIds.slice(i);
+      api.updateMessageFlags(auth, affectedIds, 'remove', 'read');
```
Hmm, yeah, thanks -- these should have `return` or `await`.

gnprice commented on src/webview/MessageList.js:
```diff
@@ -157,6 +161,7 @@ function useMessageListProps(props: OuterProps): Props {
     ),
     typingUsers: getCurrentTypingUsers(state, props.narrow),
     doNotMarkMessagesAsRead,
+    setDoNotMarkMessagesAsRead,
```
Sure.

gnprice commented on src/webview/MessageList.js:
```diff
@@ -125,21 +128,22 @@ function useMessageListProps(props: OuterProps): Props {
-          return false;
-      }
-    })();
+  const [doNotMarkMessagesAsRead, setDoNotMarkMessagesAsRead] = useState(() => {
```
Yeah, good thought.

In fact, we already weren't set up to handle `doNotMarkMessagesAsRead` changing. It's used in generating the initial HTML, where it initializes a variable of the same name that's consulted in `js.js` — and we didn't have an inbound event to update that if it changes. This is actually a bug in this version of the PR, because it means that the `setDoNotMarkMessagesAsRead(true)` in the action-sheet action fails to stop the green unread-indicators from getting erased. The next revision will fix that.

I think it may not be too complicated to write this so that it works straightforwardly if the desired initial value did somehow change. So I may just do that.

> In fact, `narrow` does change in a few cases:

(Happily, though, neither of those cases changes from one type of narrow to another; so `isConversationNarrow(narrow)` remains constant, and the initial value of `doNotMarkMessagesAsRead` isn't affected.)

chrisbobbe replied:
> (Happily, though, neither of those cases changes from one type of narrow to another; so isConversationNarrow(narrow) remains constant, and the initial value of doNotMarkMessagesAsRead isn't affected.)

True! I had reasoned through that too, but thought it might be good to put in a code comment, and also it's worth knowing whether we might one day add a case that changes the narrow type.

gnprice commented:
Thanks @chrisbobbe for the careful review!

Revision pushed. This addresses the review feedback above, and also switches to using the draft new API from https://github.com/zulip/zulip/pull/23512 .

PR is still a draft because the server side isn't merged; but other than updating for the final merged version of that (in particular if the feature level changes), I believe it's ready.

chrisbobbe commented on src/action-sheets/index.js:
```diff
@@ -179,6 +185,85 @@ const markTopicAsRead = {
+        // This should be impossible given that found_newest was false
+        // (and that our num_after was positive.)
+        logging.error('mark-unread: request failed to make progress', result);
+        throw new Error(_('The server sent a malformed response.'));
```
I notice this is translated, but I don't expect this message will ever reach the user. I think any error thrown will cause `markAsUnreadFromMessage.errorMessage` to be shown, in translation, in an error alert. That string is "Failed to mark as unread".

gnprice replied:
It shows both -- that latter string is the title of the alert, and the error's own message is the body. I saw this in manual testing with an artificial error, and here's the code in `makeButtonCallback`:
```js
      try {
        await pressedButton.action(args);
      } catch (err) {
        // […]
        Alert.alert(args._(pressedButton.errorMessage), err.message);
      }
```

chrisbobbe replied:
Oh OK, I missed that. Thanks!

chrisbobbe reviewed (commented):
Thanks! This looks good, with a few small comments below.

gnprice commented on src/api/messages/updateMessageFlagsForNarrow.js:
```diff
@@ -0,0 +1,32 @@
+  found_newest: boolean,
+|};
+
+/** https://zulip.com/api/update-message-flags-for-narrow */
+export default function updateMessageFlagsForNarrow(
+  auth: Auth,
+  args: {|
+    narrow: ApiNarrow,
+    anchor: Anchor,
+    include_anchor?: boolean,
+    num_before: number,
+    num_after: number,
+    op: 'add' | 'remove',
+    flag: UserMessageFlag,
+  |},
+): Promise<ApiResponseUpdateMessageFlagsForNarrow> {
+  return apiPost(auth, '/messages/flags/narrow', { ...args, narrow: JSON.stringify(args.narrow) });
+}
```
At a dynamic level, such a check isn't really necessary because on an older server, the request will promptly fail because the endpoint doesn't exist -- which isn't much different from having an assertion that fails.

In `api.updateUserSettings`:
```js
/**
 * https://zulip.com/api/update-settings
 *
 * NOTE: Only supports servers at feature level 89+ because UserSettings
 * assumes 89+; see there.
 */
// TODO(server-5.0): Simplify FL 89+ requirement
export default (/* … */): Promise<ApiResponseSuccess> => {
  invariant(zulipFeatureLevel >= 89, 'api.updateUserSettings not supported with FL <89');
```
the check was more necessary because the endpoint itself is older, so that it's more complicated to think about what happens if the server is older -- the endpoint exists and just behaves differently. So the assertion potentially saves us from some more complicated and puzzling behavior we'd get by making the actual request.

Because the assertion could never fire on up-to-date servers, which are what we normally develop against, it also doesn't help find bugs at development time.

There would be value in having the version requirement documented in the jsdoc. In principle it's all there in the linked API docs, though, because (again unlike `api.updateUserSettings`) the version requirement is just restating the "this endpoint was added in …" paragraph from the API docs. So I'm inclined to stick to letting the API docs speak for themselves -- in general when adding or reviewing a new call site for an API endpoint, one is going to need to consult the API docs in any case.

chrisbobbe replied:
Got it, makes sense!

gnprice commented on src/action-sheets/index.js:
```diff
@@ -620,6 +714,22 @@ export const constructMessageActionButtons = (args: {|
+    // … and it isn't to a stream we're not subscribed to.
+    && (message.type === 'private' || backgroundData.subscriptions.get(message.stream_id))
+    // Oh and if the server is too old for the feature, don't offer it.
+    // TODO(server-6): Simplify this away.
```
Yeah. I guess I'll just make it "server-6.0" for now.

gnprice commented:
Thanks for the review!

Revised for the nit above. Otherwise I think we're just waiting for the API tweaks discussed today here:
https://chat.zulip.org/#narrow/stream/378-api-design/topic/mark-as-unread.20request/near/1465782
to be done; then I'll update this to match and we can retest and hopefully declare everything ready to finalize.

gnprice commented:
OK -- zulip/zulip#23512 is merged, and all seems well in my testing, so just merged this.
